mypy/build.py

```diff
@@ -926,6 +927,84 @@ def verify_module(fscache: FileSystemMetaCache, id: str, path: str) -> bool:
+    proto_meta, proto_cache = get_protocol_deps_cache_name(manager)
+    try:
+        with open(proto_meta, 'r') as f:
+            data = f.read()
+            manager.trace('Proto meta {}'.format(data.rstrip()))
+            meta_snapshot = json.loads(data)
+    except IOError:
+        manager.log('Could not load protocol meta: could not find {}'.format(proto_meta))
+        return None
+    current_meta_snapshot = {}  # type: Dict[str, str]
+    for id in graph:
+        meta = graph[id].meta
+        assert meta is not None, 'Protocol cache should be read after all other.'
+        current_meta_snapshot[id] = meta.hash
+    common = set(meta_snapshot.keys()) & set(current_meta_snapshot.keys())
+    if any(meta_snapshot[id] != current_meta_snapshot[id] for id in common):
+        # TODO: invalidate also if options changed (like --strict-optional)?
+        manager.log('Protocol cache inconsistent, ignoring.')
+        return None
+    try:
+        with open(proto_cache, 'r') as f:
+            data = f.read()
+            manager.trace('Proto deps {}'.format(data.rstrip()))
+            deps = json.loads(data)  # TODO: Errors
+    except IOError:
+        manager.log('Could not load protocol cache: could not find {}'.format(proto_cache))
+        return None
+    if not isinstance(deps, dict):
+        manager.log('Could not load protocol cache: cache is not a dict: {}'
+                    .format(repr(deps)))
+        return None
+    return {k: set(v) for (k, v) in deps.items()}
+
+
+def collect_protocol_deps(graph: Graph) -> Dict[str, Set[str]]:
+    """Collect protocol dependency map for fine grained incremental mode."""
+    from mypy.server.deps import collect_protocol_attr_deps
+    result = {}  # type: Dict[str, Set[str]]
+    for id in graph:
+        file = graph[id].tree
+        assert file is not None, "Should call this only when all files are processed"
+        names = file.names
+        if '/typeshed/' not in file.path and id not in ('typing', 'builtins'):
+            deps = collect_protocol_attr_deps(names)
+            for trigger, targets in deps.items():
+                result.setdefault(trigger, set()).update(targets)
+    return result
+
+
+def write_protocol_deps_cache(proto_deps: Dict[str, Set[str]],
+                              manager: BuildManager, graph: Graph) -> None:
+    """Write cache files for protocol dependencies.
+
+    Serialize protocol dependencies map for fine grained mode. Also take the snapshot
+    of current sources to later check consistency between protocol cache and individual
+    cache files.
+    """
+    proto_meta, proto_cache = get_protocol_deps_cache_name(manager)
+    meta_snapshot = {}  # type: Dict[str, str]
+    for id in graph:
+        meta_snapshot[id] = graph[id].source_hash
+    if not atomic_write(proto_meta, json.dumps(meta_snapshot), '\n'):
+        manager.log("Error writing protocol meta JSON file {}".format(proto_cache))
```

Comment:
Error handling is a bit sketchy -- the cache is unusable without a protocol meta JSON file. Other cache files can be regenerated by reprocessing individual modules, but the global cache can't be built incrementally. So if we write some cache files but not the global cache file we should probably fail the build. If we don't write any cache files at all, things are probably okay, since we'll need to a full rebuild which will also populate the protocol data structures correctly.

Reply:
Yes, this is the part where I am not sure what to do. Also should we invalidate the protocol cache separately if cache was written with `--strict-optional` but is used without `--strict-optional`?